File: frontend/app/types/reservation.ts
export interface PagedResult<T> {
  items: T[]
  total: number
  page: number
  pageSize: number
}

export type ReservationStatus = 'New' | 'Consulting' | 'Confirmed' | 'Visited' | 'Cancelled'

export interface ReservationListItem {
  id: number
  code: string
  name: string
  wechatId: string
  status: ReservationStatus
  consultantId: number | null
  consultantName: string | null
  createdAt: string
  visitDate: string | null
}

export interface ReservationSummary {
  new: number
  consulting: number
  confirmed: number
  visitedThisMonth: number
}

export interface ReservationNote {
  id: number
  body: string
  authorUserId: number | null
  authorName: string
  createdAt: string
  updatedAt: string
  isEdited: boolean
}

export interface ReservationLog {
  id: number
  action: string
  note: string | null
  actorName: string
  createdAt: string
}

export interface ReservationNoteRevision {
  id: number
  body: string
  editedByName: string
  editedAt: string
}

export interface ReservationDetail {
  id: number
  code: string
  name: string
  birthDate: string
  gender: 'Female' | 'Male' | 'Other'
  wechatId: string
  preferredContactDate: string | null
  preferredContactTime: string | null
  locale: string
  status: ReservationStatus
  consultantId: number | null
  consultantName: string | null
  visitDate: string | null
  visitTime: string | null
  depositAmount: number | null
  depositCurrency: 'CNY' | 'KRW'
  depositPaid: boolean
  cancelReason: string | null
  utmSource: string
  utmMedium: string
  utmCampaign: string
  referralCode: string
  createdAt: string
  updatedAt: string
  consultingAt: string | null
  confirmedAt: string | null
  visitedAt: string | null
  cancelledAt: string | null
  procedureIds: number[]
  notes: ReservationNote[]
  logs: ReservationLog[]
}

export interface ReservationCalendarItem {
  id: number
  code: string
  name: string
  status: ReservationStatus
  visitDate: string
  visitTime: string | null
  consultantName: string | null
}

// 달력 그리드 배지용 — 날짜별 건수만(2026-08-27, 상세 목록과 분리)
export interface ReservationCalendarDayCount {
  visitDate: string
  count: number
}

export interface ConsultantLookup {
  id: number
  name: string
  isActive: boolean
  sortOrder: number
}

// D25(2026-08-28) — sortOrder 폐지, categoryId 추가(예약 상세 아코디언이 이 값으로 카테고리별 그룹핑).
export interface ProcedureLookup {
  id: number
  code: string
  categoryId: number
  nameZhCn: string
  nameZhTw: string
  nameEn: string
  nameKo: string
  isActive: boolean
}

// D25 — 시술 카테고리 마스터. 관리 탭 목록 + 예약 상세 시술 아코디언 그룹 헤더가 공유.
export interface CategoryLookup {
  id: number
  code: string
  nameZhCn: string
  nameZhTw: string
  nameEn: string
  nameKo: string
  isActive: boolean
}

// Phase 6 — 실장 KPI(11-4절). 활성 실장은 실적 0건이어도 0행으로 내려온다(11-6절 구간 0 채움).
export interface ConsultantKpi {
  consultantId: number
  consultantName: string
  assigned: number
  confirmed: number
  visited: number
  conversionRate: number
}

// Phase 6 — 예약 통계 주간 추이(D16). weekStart는 date(타임존 없음) 컬럼이라 KST 변환 불필요.
export interface WeeklyReservationStat {
  weekStart: string
  received: number
  confirmed: number
  visited: number
  cancelled: number
}

export interface ProcedureStat {
  procedureId: number
  nameZhCn: string
  nameZhTw: string
  nameEn: string
  nameKo: string
  count: number
}

export interface LocaleStat {
  locale: string
  count: number
}

// 담당 실장 축(11-4절) — 비활성 실장 제외, KPI와 달리 0행 채움 없음(실적 있는 실장만).
export interface ConsultantReservationStat {
  consultantId: number
  consultantName: string
  count: number
}

export interface ReservationStats {
  weekly: WeeklyReservationStat[]
  procedures: ProcedureStat[]
  locales: LocaleStat[]
  consultants: ConsultantReservationStat[]
}

// Phase 8 — 유입 경로 분석(D4·D5, 15-2절). 방문 기록(landing_daily_stats)이 있는 조합만 내려온다.
export interface ReferralStat {
  referralCode: string
  utmSource: string
  utmMedium: string
  utmCampaign: string
  visitCount: number
  reservationCount: number
  conversionRate: number
  confirmedCount: number
  confirmedConversionRate: number
}

// B안, 2026-08-27 신설 — 인플루언서 짧은 링크(/go/{code}) 매핑. code는 생성 후 변경 불가.
export interface InfluencerLink {
  id: number
  code: string
  displayName: string
  utmSource: string
  utmMedium: string
  utmCampaign: string
  locale: string
  isActive: boolean
  createdAt: string
}

export type AdminRole = 'Admin' | 'HospitalManager' | 'Consultant'

export interface AdminUser {
  id: number
  email: string
  role: AdminRole
  name: string
  locale: string
  isSuspended: boolean
  createdAt: string
}

export interface AuditLogEntry {
  id: number
  actorUserId: number | null
  actorEmail: string
  actorRole: string
  action: string
  entityType: string
  entityId: string | null
  summary: string
  ip: string | null
  statusCode: number
  createdAt: string
}

File: api/Models/Reservation.cs
namespace WonjinApi.Models;

// 핵심 테이블. 소프트 삭제는 AppDbContext의 전역 쿼리 필터(deleted_at IS NULL)로 강제한다(D15).
public class Reservation
{
    public int Id { get; set; }
    public string Code { get; set; } = string.Empty; // YYYYMMDD+4자리, 8-11절 카운터로 발급
    public string Name { get; set; } = string.Empty;
    public DateOnly BirthDate { get; set; }
    public string Gender { get; set; } = string.Empty; // Female | Male | Other
    public string WechatId { get; set; } = string.Empty;
    // 고객 직접 입력, KST 벽시계 날짜·시각 (D10). Date는 2026-08-28 추가 — 라이브 서비스 기존 행
    // 호환을 위해 NULL 허용이되 신규 제출은 프론트·백엔드 모두 필수(visit_date와 동일 취급).
    // 🔴 D26(2026-08-28) — "상관없음" 체크 시 둘 다 NULL로 저장(시각 무관 표시).
    public DateOnly? PreferredContactDate { get; set; }
    public TimeOnly? PreferredContactTime { get; set; }
    public string Locale { get; set; } = string.Empty;
    public string Status { get; set; } = "New"; // New | Consulting | Confirmed | Visited | Cancelled

    public int? ConsultantId { get; set; } // NULL이면 미배정 — D17: 배정 전엔 업무 입력 전부 차단
    public Consultant? Consultant { get; set; }

    public DateOnly? VisitDate { get; set; }
    public TimeOnly? VisitTime { get; set; }
    public decimal? DepositAmount { get; set; }
    public string DepositCurrency { get; set; } = "CNY"; // CNY | KRW, 환율 환산 안 함(D12)
    public bool DepositPaid { get; set; }
    public string? CancelReason { get; set; }

    public string UtmSource { get; set; } = string.Empty;
    public string UtmMedium { get; set; } = string.Empty;
    public string UtmCampaign { get; set; } = string.Empty;
    public string ReferralCode { get; set; } = string.Empty;

    public DateTimeOffset CreatedAt { get; set; }
    public DateTimeOffset UpdatedAt { get; set; }
    public DateTimeOffset? ConsultingAt { get; set; }
    public DateTimeOffset? ConfirmedAt { get; set; }
    public DateTimeOffset? VisitedAt { get; set; }
    public DateTimeOffset? CancelledAt { get; set; }

    public DateTimeOffset? DeletedAt { get; set; } // D15 소프트 삭제
    public int? DeletedByUserId { get; set; }
    public User? DeletedByUser { get; set; }

    public ICollection<ReservationProcedure> ReservationProcedures { get; set; } = [];
    public ICollection<ReservationNote> Notes { get; set; } = [];
    public ICollection<ReservationLog> Logs { get; set; } = [];

    // 낙관적 동시성 토큰 — PostgreSQL 시스템 컬럼 xmin에 매핑(AppDbContext에서 IsRowVersion() 설정).
    // 새 컬럼을 추가하는 게 아니라 이미 모든 행에 존재하는 값을 노출만 하는 것이라 저장공간 증가 없음.
    public uint RowVersion { get; set; }
}
